### PR TITLE
GroupConversation: open and centralize some boolean flags.

### PR DESCRIPTION
**Pull request by abourget**
In https://github.com/abourget/slick .. I keep references to those channels and conversations, and I sometimes tweak them... I need to be able to create some and keep them updated..

any reasons to keep those internal ?

This is just a proposition.. nothing fixed in stone here..

### CODE REVIEW

**Comment by abourget** [comment]
so for simplicty of the bot, I like to have a "Channel" being able to represent a Group.

Why have `IsChannel` on a Channel.. if there's no way it is not a channel.. why have `IsGroup` on a Group, if there's no way it's not a group.

I feel we should have a `Channel` with "IsGroup" and distinguish it like that.. shouldn't we ? would be way simpler..

**Comment by abourget** [comment]
@mlbvsnba was that unified Channel model something that you found useful in my fork ?

**Comment by abourget** [comment]
Ok, this last commit re-similarizes Group and Channel.. through `GroupConversation`.

**Comment by abourget** [comment]
Cool.. I just re-migrated to your breaking changes branch.. couple of quick fixes and I'm back on track (with this PR in).

Thanks for merging all of this !

**Comment by abourget** [comment]
Whoops, added yet another feature. Support for Websockets through HTTP_PROXYs !

**Comment by nlopes** [comment]
I agree with the channel/group changes as I thought of changing it myself to make things easier. The websocket proxy change maintains backwards compatibility so I am good with it as well. Thanks!

File: im.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ type imResponseFull struct {
 
 // IM contains information related to the Direct Message channel
 type IM struct {
-	conversation
+	Conversation
```
Comment (joefitzgerald) [inline im.go:25]: This shouldn't be exported. The fields within conversation are exported and can be accessed that way.

File: groups.go [hunk 2]
```diff
@@ -8,8 +8,7 @@ import (
 
 // Group contains all the information for a group
 type Group struct {
-	groupConversation
-	IsGroup bool `json:"is_group"`
+	GroupConversation
```
Comment (joefitzgerald) [inline groups.go:11]: I think you may have missed the point of adding GroupConversation. It includes the common members that are shared by both Channel and Group. You cannot be a member (`IsMember`) of a Channel. You cannot have a group that `IsGeneral`. `IsChannel` and `IsGroup` are mutually exclusive. Channels and Groups never come back in a unified list. Consequently, I'm not sure why you made the change?

Further, groupConversation shouldn't be exported.

File: conversation.go [hunk 3]
```diff
@@ -12,10 +12,14 @@ type conversation struct {
 }
 
 // GroupConversation is the foundation for Group and Channel
-type groupConversation struct {
-	conversation
+type GroupConversation struct {
+	Conversation
 	Name       string   `json:"name"`
 	Creator    string   `json:"creator"`
+	IsChannel  bool     `json:"is_channel"`
```
Comment (joefitzgerald) [inline conversation.go:19]: Not valid in a Group.

File: conversation.go [hunk 4]
```diff
@@ -12,10 +12,14 @@ type conversation struct {
...
+	IsGroup    bool     `json:"is_group"`
```
Comment (joefitzgerald) [inline conversation.go:20]: Not valid in a Channel.

File: conversation.go [hunk 5]
```diff
@@ -12,10 +12,14 @@ type conversation struct {
...
+	IsGeneral  bool     `json:"is_general"`
```
Comment (joefitzgerald) [inline conversation.go:21]: Not valid in a Group.

File: conversation.go [hunk 6]
```diff
@@ -12,10 +12,14 @@ type conversation struct {
...
+	IsMember   bool     `json:"is_member"`
```
Comment (joefitzgerald) [inline conversation.go:22]: Not valid in a Channel.

File: channels.go [hunk 7]
```diff
@@ -18,10 +18,7 @@ type channelResponseFull struct {
 
 // Channel contains information about the channel
 type Channel struct {
-	groupConversation
-	IsChannel bool `json:"is_channel"`
-	IsGeneral bool `json:"is_general"`
-	IsMember  bool `json:"is_member"`
+	GroupConversation
```
Comment (joefitzgerald) [inline channels.go:21]: I think you may have missed the point of adding GroupConversation. It includes the common members that are shared by both Channel and Group. You cannot be a member (`IsMember`) of a Channel. You cannot have a group that `IsGeneral`. `IsChannel` and `IsGroup` are mutually exclusive. Channels and Groups never come back in a unified list. Consequently, I'm not sure why you made the change?

Further, groupConversation shouldn't be exported.

File: conversation.go [hunk 8]
```diff
@@ -1,7 +1,7 @@
 package slack
 
 // Conversation is the foundation for IM and BaseGroupConversation
-type conversation struct {
```
Comment (joefitzgerald) [inline conversation.go:4]: This shouldn't be exported. The fields within conversation are exported and can be accessed when embedded.

File: conversation.go [hunk 9]
```diff
@@ -12,10 +12,14 @@ type conversation struct {
 }
 
 // GroupConversation is the foundation for Group and Channel
-type groupConversation struct {
-	conversation
+type GroupConversation struct {
```
Comment (joefitzgerald) [inline conversation.go:15]: This shouldn't be exported. The fields within groupConversation are exported and can be accessed when embedded.

File: conversation.go [hunk 10]
```diff
@@ -12,10 +12,14 @@ type conversation struct {
...
+	Conversation
```
Comment (joefitzgerald) [inline conversation.go:16]: This shouldn't be exported. The fields within conversation are exported and can be accessed that way.

**Comment by joefitzgerald** [comment]
@abourget I'm trying to figure out why you wanted these changes? Is it simply because you want to represent a single list of all channels and groups and be able to perform actions based on that list and the type of the group conversation? These changes create JSON that is not valid when interacting with `channel` and `group` resources.

**Comment by abourget** [comment]
Ah I see. Yes the goal is to work with a unified list. It simplifies a lot
of things in a statically typed world. When reading having "IsGroup" on a
Group doesn't provide much value.. where it could never be false. However
having IsGroup being false on a channel is helpful.

I understand that when writing, things are different. Could we make these
precise fields be pointers to bool, and add "omitempty" for when we write ?
Does the API fail when we send an `"is_group": false` in a channel context ?

Also regarding the Conversation being exported, if it isn't,  you can't
create those objects.. only modify its fields.. so you need to have huge
copy functions.  So I ask you, why not export them ?

Le lun. 31 août 2015 17:39, Joe Fitzgerald notifications@github.com a
écrit :

**Comment by nlopes** [comment]
`Does the API fail when we send an`"is_group": false`in a channel context ?` - I am going to test this because this is something I was thinking as well.
On the other hand, I really don't like giving the ability to people to set IsGroup to true on a Channel by mistake. @joefitzgerald brings some valid points to the conversation and they're probably a bit saner (from a developer perspective) so I am going to re-analyse this merge and might revert it.

**Comment by abourget** [comment]
In the same vein, I need ReactionEvent to be opened up.. for the same
reason. I want to be able to deal with both reactions with a single API.

Le jeu. 10 sept. 2015 14:48, Norberto Lopes notifications@github.com a
écrit :

**Comment by abourget** [comment]
In which case do people create Channel objects ? Those objects are read into.. and in that case theirs fields are appropriate with my changes. We always refer to channels with their ID in API calls.

When I create Channel objects.. it is to avoid duplication of my code, by considering a Group as a Channel, which it _is_ conceptually  (and most probably in Slack's own implementation!).

I'm sorry to say, I don't see the point being made by "groupConversation shouldn't be exported". I'm presenting a use case, "it shouldn't" is not a compelling reason enough to dismiss it.

Again, having IsChannel and IsGroup is absolutely useless in Go if we only have it on their respective objects, because the _type_ of the object already tells us that. IsChannel and IsGroup is only useful if it can discriminate _against_ something else, which it can help do in a GroupConversation.

I'd prefer not to maintain my own fork because of issues like this. Can we settle on this matter ? Especially in the light of a similar change regarding ReactionEvent I'm about to propose.

**Comment by abourget** [comment]
Ok folks, I'm willing to revert the groupConversation changes if you want.. I needed to have the same support for `IM`and `IsIM` but making it clean in the `slack` codebase was becoming very ugly.. so I reverted to create my own abstraction in my bot.. This means I wouldn't need the GroupConversation to be open (as I'm not going to use the `slack.Channel` for the abstraction), it can be closed up, and we can put back the `IsChannel` only on the `Channel` object.  Just beware that this PR included the HTTP_PROXY support too.. so maybe we want to revert only the two concerned commits.

**Comment by nlopes** [comment]
Awesome, thank you for this. I'll be reverting that either today or tomorrow.

**Comment by nlopes** [comment]
Argh, I ended up spending more time away from this than I wished.
The two commits have been reverted, the change for HTTP_PROXY has been left in.

Thank you so much @abourget

**Pull request merged**